### PR TITLE
External etcd to support HA setup and scaling with etcd

### LINKED ISSUE #479: Support HA control-plane with etcd
Hi, I just wanted to ask if there are any plans to support HA control-plane with using the internal etcd of k0s? I don't quite understand the reasoning in the docs:

> As the nature of Kubernetes workloads, in this case the cluster control planes, is quite dynamic it poses a challenge to setup highly available Etcd cluster for the control plane.
Source: https://docs.k0smotron.io/stable/ha/

With a StatefulSet and robust PVC storage, I don't really see any blocker to run an etcd cluster within k8s and it would actually simplify the HA setup quite a bit. What is currently missing to support this, is it a limitation of k0s or just the way k0smotron provisions clusters?

### PR DESCRIPTION
**Pull request by makhov**
Fixes #479

We decided to stick with the external etcd statefulset for hosted control planes for a couple of reasons:
- It allows easy creation of HA clusters with etcd
- It supports scaling up
- It simplifies etcd cluster maintenance

### CODE REVIEW

File: internal/controller/k0smotron.io/k0smotroncluster_etcd.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+			},
+		},
+	}
+
+	_ = ctrl.SetControllerReference(kmc, &svc, r.Scheme)
+
+	return r.Client.Patch(ctx, &svc, client.Apply, patchOpts...)
+}
+
+func (r *ClusterReconciler) reconcileEtcdStatefulSet(ctx context.Context, kmc *km.Cluster) error {
```
Comment (jnummelin) [inline internal/controller/k0smotron.io/k0smotroncluster_etcd.go:91]: The logic is bit hard to follow, maybe few comments would be in order.

Also if we'd separate the "calculation" into separate func we could also have proper unit tests for it. Tests also makes it easier to understand, IMO.

File: internal/controller/k0smotron.io/k0smotroncluster_controller.go [hunk 2]
```diff
@@ -135,8 +135,29 @@ func (r *ClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
...
+			return ctrl.Result{Requeue: true, RequeueAfter: time.Minute}, fmt.Errorf("error generating etcd certificates: %w", err)
+		}
+		kmc.Spec.CertificateRefs = append(kmc.Spec.CertificateRefs, km.CertificateRef{
+			Type: string(secret.APIServerEtcdClient),
+			Name: secret.Name(kmc.Name, secret.APIServerEtcdClient),
+		})
+	}
+
+	logger.Info("Reconciling etcd")
+	if err := r.reconcileEtcd(ctx, &kmc); err != nil {
```
[the remark this answers is not in the document]
Comment (makhov) [reply internal/controller/k0smotron.io/k0smotroncluster_controller.go:157]: It is guarded, at the begining of the `reconcileEtcd` function.

File: inttest/capi-controlplane-docker-downscaling/capi_controlplane_docker_downscaling_test.go [hunk 3]
```diff
@@ -152,6 +152,8 @@ func (s *CAPIControlPlaneDockerDownScalingSuite) TestCAPIControlPlaneDockerDownS
 
 	s.T().Log("scaling down control plane")
 
+	time.Sleep(5 * time.Second)
+
 	out, err := exec.Command("docker", "exec", "docker-test-1", "k0s", "etcd", "leave").CombinedOutput()
```
[the remark this answers is not in the document]
Comment (makhov) [reply inttest/capi-controlplane-docker-downscaling/capi_controlplane_docker_downscaling_test.go]: Unfortunately, due to how docker terminates containers, the script isn't executed

File: internal/controller/k0smotron.io/k0smotroncluster_etcd.go (hunk 1 above)
Comment (makhov) [reply internal/controller/k0smotron.io/k0smotroncluster_etcd.go:91]: Added some comments and tests

**Review by jnummelin — APPROVED** [review]
